cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -0,0 +1,90 @@
+        // Run each phase, gathering info along the way until we get to the end
+        // and can send it all to the catalog
+
+        // Best pattern to use here?
+        // - Store in structs in Self along the way?
+        // - Type States (like the old publish?)
+        // - Make each phase independant, taking arguments for all that's needed
+        //    and track the ongoing state only here in publish()?
+        // - Combination of Type States and making each phase independant may be the 
+        //    easiest to test?
+        prepublish_check();
+
+        prepublish_build();
+
+        gather_build_info();
+
+        publish_to_catalog();
+    }
+
+    /// Ensures current commit is in remote
+    /// Gathers of the info needed for publish
+    /// ... locked_url, rev, rev_count, rev_date of the build repo from git
+    /// ... locked_url, rev, rev_count, rev_date of the base catalog from the
+    ///         lockfile of the build repo
+    /// ... version, description, name, attr_path from manifest
+    /// ... catalog from ???? (cli argument? manifest?)
+    pub fn prepublish_check(repo_dir: &Path, package_name: &str)
+    {
+        // _sandbox = temp dir
+        // Use GitCommandProvider to get remote and current rev of build_repo
+        // Use GitCommandProvider to clone that remote/rev to _sandbox
+            // this will ensure it's in the remote
+        // Load the manifest from the .flox of that repo to get the version/description/catalog
+    }
+    
+    /// Access a `flox build` builder, probably another provider?
+    /// ... to clone into a sandbox and run
+    /// ... `flox activate; flox build;`
+    pub fn prepublish_build<B: ManifestBuilder>(builder: &B, _sandbox: &Path)
+    {
+        // We need to do enough to get the build info for the next step
+        // ... equivalent to run `flox activate; flox build;`?
+
+        // This will create a flox env in the _sandbox, right? so should we pass that back
+        // and store it to use for the next step?
+
+        // Use ManifestBuilder to perform the build (evaluation?) so we can get
+        // the info we need.  Do we need to build? or is there something else?
+        let build_output = builder.build(_sandbox, flox_env, &self.package_name);
+        
+        // the build_output looks to be the stdout of the build process.. how do 
+        // we access the drv_path and that stuff?
+    }
+
+    /// Obtains info from the build process (the sandboxed one above)
+    /// ... base_catalog_{locked_url, rev, rev_count, rev_date}
+    /// ... system, drv_path (??), outputs (??)
+    pub fn gather_build_info(_sandbox: &Path)
+    {
+        // Access lockfile of _sandbox to get base_catalog_{locked_url, rev, rev_count, rev_date}
+        // populate the following:
+        //      how do we get the drv_path, outputs(and paths), system from this?
+    }
+
+    /// Uses client catalog to...
+    /// ... check access to the catalog
```

Comment:
We should probably check access to the catalog earlier so that we can avoid a potentially expensive build if the user doesn't have access.

Reply:
Noted, thanks.